tests/tensorflow/nightly/common.libsonnet

```diff
@@ -25,6 +28,174 @@ local mixins = import 'templates/mixins.libsonnet';
       softwareVersion: 'nightly',
     },
     imageTag: 'nightly',
+    podTemplate+:: if config.accelerator.type == 'tpu' then
+      {
+        spec+: {
+          initContainerMap+:: {
+            'tpu-version': {
+              image: config.podTemplate.spec.containerMap.train.image,
+              env+: [
+                {
+                  name: 'TPU_NAME',
+                  valueFrom: {
+                    fieldRef: {
+                      fieldPath: "metadata.annotations['name.cloud-tpus.google.com/train']",
+                    },
+                  },
+                },
+                {
+                  name: 'POD_UID',
+                  valueFrom: {
+                    fieldRef: {
+                      fieldPath: 'metadata.uid',
+                    },
+                  },
+                },
+              ],
+              local tpuCreateSettings = {
+                acceleratorName: std.escapeStringBash(config.accelerator.name),
+                softwareVersion: std.escapeStringBash(config.tpuSettings.softwareVersion),
+                startupScript: std.escapeStringBash(config.tpuSettings.tpuVmStartupScript),
+                sleepTime: config.tpuSettings.tpuVmCreateSleepSeconds,
+                testName: std.strReplace(config.testName, '.', '-'),
+              },
+              command: [
+                'python3',
+                '-c',
+                |||
+                  import os
+                  import tensorflow as tf
+                  import urllib
+                  import json
+                  import cloud_tpu_client
+                  import sys
+                  print('python version: ' + str(sys.version))
+                  print('tf_version: ' + str(tf.__version__))
+                  #TODO(chandrasekhard):
+                  # Add extra condition to fail if it picks stale image
+                  print(str(tf.__file__))
+                  ctc = cloud_tpu_client.Client(tpu=os.path.basename('$(TPU_NAME)'), zone=os.path.dirname('$(TPU_NAME)'))
+                  ctc.wait_for_healthy()
+                  ctc.configure_tpu_version('nightly', restart_type='always')
+                  ctc.wait_for_healthy()
+                  _VERSION_SWITCHER_ENDPOINT = 'http://{}:8475/requestversion'
+                  url = _VERSION_SWITCHER_ENDPOINT.format(ctc.network_endpoints()[0]['ipAddress'])
+                  req = urllib.request.Request(url)
+                  resp = urllib.request.urlopen(req)
+                  version_details = json.loads(resp.read())
+                  print(version_details)
+                |||,
+              ],
+            },
+          },
+        },
+      }
+    else
+      {},
+  },
+  tpuVm:: experimental.TensorFlowTpuVmMixin {
```

Comment:
Sorry I missed this on my last pass: you should be able to eliminate all fo the overrides here except for `podTemplate.spec.initContainerMap['create-tpu'].command`, since that's the only thing that's changing from the parent.

Reply:
Actually I tried that but the local variables like tpuCreateSettings is not reflected here.  There are other unbound variable errors that I encountered.

Reply:
You won't be able to take out the `local` variables if they're using in `command` like `tpuCreateSettings`, since they're scoped to this object. You can take out e.g. `podTemplate.spec.initContainerMap['create-tpu'].image` since that will be inherited

Reply:
Removing that line results in an error - 

The Job "tf.nightly-maskrcnn-coco-func-v4-8-1vm-jbrz7" is invalid: spec.template.spec.initContainers[0].image: Required value

Reply:
I see use `'create-tpu'+:` (note the `+`) to override individual fields rather than replace all of them.

Reply:
Done